File: solutions/javascript/complex-numbers/1/complex-numbers.js

export class ComplexNumber {
  
  constructor(real, imag) {
    this._real = real;
    this._imag = imag;
  }

  get real() {
    return this._real;
  }

  get imag() {
    return this._imag;
  }

  add(other) {
    return new ComplexNumber(this.real + other.real, this.imag + other.imag);
  }

  sub(other) {
    return new ComplexNumber(this.real - other.real, this.imag - other.imag);
  }

  div(other) {
    const denum = (other.real * other.real + other.imag * other.imag);
    const real = (this.real * other.real + this.imag * other.imag) / denum;
    const imag = (this.imag * other.real - this.real * other.imag) / denum;
    return new ComplexNumber(real, imag);
  }

  mul(other) {
    const real = this.real * other.real - this.imag * other.imag;
    const imag = this.real * other.imag + this.imag * other.real;
    return new ComplexNumber(real, imag);
  }

  get abs() {
    return Math.sqrt(this.real * this.real + this.imag * this.imag);
  }

  get conj() {
    // javascript supports 0 and -0, we always want 0 here
    const imag = this.imag == 0 ? 0 : -this.imag;
    return new ComplexNumber(this.real, imag);
  }

  get exp() {
    const expReal = Math.exp(this.real);
    return new ComplexNumber(expReal * Math.cos(this.imag), expReal * Math.sin(this.imag));
  }
}

File: solutions/go/complex-numbers/1/complex_numbers.go
package complexnumbers

import "math"

// Define the Number type here.
type Number struct {
    real float64
    imag float64
}

func (n Number) Real() float64 {
	return n.real
}

func (n Number) Imaginary() float64 {
	return n.imag
}

func (n1 Number) Add(n2 Number) Number {
	return Number{n1.real + n2.real, n1.imag + n2.imag}
}

func (n1 Number) Subtract(n2 Number) Number {
	return Number{n1.real - n2.real, n1.imag - n2.imag}
}

func (n1 Number) Multiply(n2 Number) Number {
	return Number{n1.real * n2.real - n1.imag * n2.imag,
                 n1.real * n2.imag + n1.imag * n2.real}
}

func (n Number) Times(factor float64) Number {
	return Number{factor * n.real, factor * n.imag}
}

func (n1 Number) Divide(n2 Number) Number {
	denom := n2.real * n2.real + n2.imag * n2.imag
    return Number{(n1.real * n2.real + n1.imag * n2.imag) /denom,
                 (-n1.real * n2.imag + n1.imag * n2.real) / denom}
}

func (n Number) Conjugate() Number {
	return Number{n.real, -n.imag}
}

func (n Number) Abs() float64 {
	return math.Sqrt(n.real * n.real + n.imag * n.imag)
}

func (n Number) Exp() Number {
    expReal := math.Exp(n.real)
	return Number{expReal * math.Cos(n.imag), expReal * math.Sin(n.imag)}
}
